src/OpenTelemetry.Extensions.Enrichment.AspNetCore/BaseAspNetCoreEnricher.cs

```diff
@@ -0,0 +1,35 @@
+// <copyright file="BaseAspNetCoreEnricher.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace OpenTelemetry.Extensions.Enrichment.AspNetCore;
+
+public abstract class BaseAspNetCoreEnricher<T>
+{
+    public virtual void EnrichWithHttpRequest(T enrichmentBag, HttpRequest request)
```

Comment:
this is same/similar to what the asp.net core instrumentation already provided. what additional value we get by doing this vs doing it the way offered by asp.net instrumentaion library? 

How do we explain to user about 2 options to do the same thing?

Reply:
Instead of delegates, users can have full-fledged enricher classes, with DI support if necessary. Then, it aligns API across all telemetry signals (log, metric, trace) and makes the API uniform and unified.

As for the 2 options - we can discuss it internally. One option is to deprecate the existing API with delegates.

cc @dpk83

Reply:
There are no plans to deprecate the existing Enrich* from the instrumentations today, so this needs to be discussed separately. We certainly do not want to  confuse users with 2 ways to achieve the same goal, unless there is significant advantage of using one over another.

The Enrich* APIs in asp.net core was intentionally added as opposed to asking users to use ActivityProcessors (the spec recommended extensibility), because Enrich* provided easy access to context/raw object which are otherwise difficult to obtain.

Reply:
The values are 
- Consistent way of enriching telemetry for customers (Be it Logs, metrics and traces)
- Ability for other libraries to configure enrichment that services can use directly. To explain further currently Enrich method can be hooked while configuring the options and there can only be one callback for enrichment in OTel. So, only the service owner has the control to hook into the enrichment callback directly. Now if a services is using 50 different libraries and some of them need to enrich telemetry then these libraries will need to expose their own mechanisms to let the services get the required data from library and then hook it up in their enrichment callback hook. Moreover each library will now implement their own mechanism to expose these properties. With this new model a library can directly add their enricher.

Also, while this PR is only adding Trace (to start small as suggested by Blanc), you should look at the overall bigger picture of Meter, Trace and Logs having the same model and then the instrumentation for trace, meter and logs having the same model. This simplifies the user experience and onboarding time greatly.

Reply:
> There are no plans to deprecate the existing Enrich* from the instrumentations today, so this needs to be discussed separately. We certainly do not want to confuse users with 2 ways to achieve the same goal, unless there is significant advantage of using one over another.

Just because we want to maintain the compatibility and not break existing API doesn't mean that we shouldn't improve the experience by providing a better mechanism in addition.

Reply:
Resolving as AspNetCore enrichment is removed from this PR